File: backend/app/api/routes/ai.py
"""AnnaSetu AI Assistance, OCR & Document Intelligence Endpoints."""

from typing import Optional
from fastapi import APIRouter, Depends, Query, status
from app.ai.schemas import (
    DocumentConsistencyResponse,
    DocumentExtractionRequest,
    DocumentExtractionResult,
    DonationAIAssistResponse,
    ExplanationResponse,
    FoodNormalizationRequest,
    FoodNormalizationResponse,
    NeedAIAssistRequest,
    NeedAIAssistResponse,
)
from app.ai.service import AIService
from app.api.deps import get_ai_service, get_current_profile
from app.schemas.auth import ProfileResponse

router = APIRouter(tags=["AI Assistance & Document Intelligence"])


@router.post(
    "/ai/normalize-food",
    response_model=FoodNormalizationResponse,
    summary="Normalize Food Description",
    description="Normalizes free-text food descriptions into candidate structured fields using deterministic parsing first, with AI assistance fallback.",
)
async def normalize_food_description(
    payload: FoodNormalizationRequest,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> FoodNormalizationResponse:
    return await ai_service.normalize_food_description(
        raw_description=payload.raw_description,
        user_id=current_profile.id,
    )


@router.post(
    "/donations/{donation_id}/ai-assist",
    response_model=DonationAIAssistResponse,
    summary="AI Assistance for Donation",
    description="Provides suggestions for category, diet, meal period, and allergen notes. Donor-declared quantity remains authoritative.",
)
async def assist_donation(
    donation_id: str,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> DonationAIAssistResponse:
    return await ai_service.assist_donation(
        donation_id=donation_id,
        user_id=current_profile.id,
    )


@router.post(
    "/needs/{need_id}/ai-assist",
    response_model=NeedAIAssistResponse,
    summary="AI Assistance for Need",
    description="Provides candidate suggestions from free-text need descriptions. Must be explicitly confirmed by NGO receiver.",
)
async def assist_need(
    need_id: str,
    payload: Optional[NeedAIAssistRequest] = None,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> NeedAIAssistResponse:
    free_text = payload.free_text_requirement if payload else None
    return await ai_service.assist_need(
        need_id=need_id,
        user_id=current_profile.id,
        free_text_requirement=free_text,
    )


@router.get(
    "/matches/{match_id}/explanation",
    response_model=ExplanationResponse,
    summary="Human-Readable Match Explanation",
    description="Generates transparent natural language explanation of deterministic priority breakdown. Algorithm score is authoritative and immutable.",
)
async def get_match_explanation(
    match_id: str,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> ExplanationResponse:
    return await ai_service.get_match_explanation(
        match_id=match_id,
        user_id=current_profile.id,
    )


@router.get(
    "/deliveries/{delivery_id}/explanation",
    response_model=ExplanationResponse,
    summary="Delivery Status Operational Explanation",
    description="Generates concise human-readable explanation based strictly on authoritative delivery state and stop progress.",
)
async def get_delivery_explanation(
    delivery_id: str,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> ExplanationResponse:
    return await ai_service.get_delivery_explanation(
        delivery_id=delivery_id,
        user_id=current_profile.id,
    )


@router.post(
    "/documents/{document_id}/process-ocr",
    response_model=DocumentExtractionResult,
    summary="Process Document OCR and Extraction",
    description="Extracts candidate fields from verification document text/image. Extracted fields are stored in verification_documents.extracted_data.",
)
async def process_document_ocr(
    document_id: str,
    payload: Optional[DocumentExtractionRequest] = None,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> DocumentExtractionResult:
    raw_ocr = payload.ocr_text if payload else None
    hint = payload.document_type if payload else None
    return await ai_service.process_verification_document(
        document_id=document_id,
        user_id=current_profile.id,
        raw_ocr_text=raw_ocr,
        doc_type_hint=hint,
    )


@router.get(
    "/documents/{document_id}/consistency-check",
    response_model=DocumentConsistencyResponse,
    summary="Check Document Consistency",
    description="Performs deterministic consistency checks between extracted document fields and user/entity profiles. Sets recommendation for manual review.",
)
async def check_document_consistency(
    document_id: str,
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> DocumentConsistencyResponse:
    return await ai_service.check_document_consistency(
        document_id=document_id,
        user_id=current_profile.id,
    )


@router.get(
    "/impact/{entity_id}/explanation",
    response_model=ExplanationResponse,
    summary="Impact Metric Explanation",
    description="Explains environmental and social impact metrics grounded solely in recorded figures and deterministic factors.",
)
async def get_impact_explanation(
    entity_id: str,
    donation_id: Optional[str] = Query(default=None, description="Optional donation ID for specific impact."),
    current_profile: ProfileResponse = Depends(get_current_profile),
    ai_service: AIService = Depends(get_ai_service),
) -> ExplanationResponse:
    return await ai_service.get_impact_explanation(
        user_id=current_profile.id,
        donation_id=donation_id,
    )

File: anna-setu-frontend-development/lib/api/ai.ts
/**
 * AnnaSetu — AI Assistance, Document OCR, and Operational Explanations API Client (Phase 16)
 *
 * IMPORTANT:
 * - AI results are suggestions only ("AI suggestion", "Extracted information", "Needs review").
 * - Deterministic backend rules remain authoritative for eligibility, matching, allocation, routing, pricing, payment, and verification.
 * - AI never independently approves or rejects verification, donations, or operations.
 */

import { apiClient } from './client'

export const AI_UI_LABELS = {
  SUGGESTION: 'AI suggestion',
  EXTRACTED_INFO: 'Extracted information',
  NEEDS_REVIEW: 'Needs review',
} as const

export interface ExtractedDocumentFields {
  document_type?: string | null
  document_number?: string | null
  holder_name?: string | null
  organization_name?: string | null
  vehicle_number?: string | null
  issue_date?: string | null
  expiry_date?: string | null
  issuing_authority?: string | null
  raw_snippet?: string | null
}

export interface DocumentExtractionResult {
  document_id: string
  document_type: string
  fields: ExtractedDocumentFields
  confidence: number
  uncertainties: string[]
  method: 'DETERMINISTIC' | 'AI_ASSISTED' | 'FALLBACK'
  review_status: 'EXTRACTED_SUCCESSFULLY' | 'EXTRACTION_UNCERTAIN' | 'DOCUMENT_MISMATCH' | 'EXPIRED_DOCUMENT' | 'MANUAL_REVIEW_REQUIRED'
  stored_in_db: boolean
}

export interface ConsistencyCheckDetail {
  field_name: string
  extracted_value?: string | null
  profile_value?: string | null
  status: 'MATCH' | 'MISMATCH' | 'PARTIAL_MATCH' | 'NOT_CHECKED'
  reason: string
}

export interface DocumentConsistencyResponse {
  document_id: string
  overall_status: 'MATCH' | 'MISMATCH' | 'PARTIAL_MATCH' | 'NOT_CHECKED'
  is_expired: boolean
  review_recommendation: string
  details: ConsistencyCheckDetail[]
  audit_note: string
}

export interface FoodNormalizationResponse {
  raw_description: string
  normalized_description: string
  possible_category?: string | null
  possible_diet_type?: string | null
  estimated_quantity_kg?: number | null
  meal_period?: string | null
  possible_allergens: string[]
  uncertainties: string[]
  is_ai_generated: boolean
}

export interface DonationAIAssistResponse {
  donation_id: string
  declared_quantity_kg: number
  normalized_description?: string | null
  category_suggestion?: string | null
  diet_suggestion?: string | null
  meal_period_suggestion?: string | null
  quantity_consistency_signal: 'CONSISTENT' | 'DISCREPANCY' | 'NOT_ESTIMATED'
  estimated_quantity_kg?: number | null
  missing_information_suggestions: string[]
  allergen_suggestions: string[]
  disclaimer: string
}

export interface NeedAIAssistResponse {
  need_id: string
  suggested_meal_period?: string | null
  suggested_diet_type?: string | null
  suggested_quantity_kg?: number | null
  special_requirements_suggestions: string[]
  uncertainties: string[]
  disclaimer: string
}

export interface ExplanationResponse {
  entity_id: string
  explanation_type: string
  explanation_text: string
  is_ai_generated: boolean
  facts: Record<string, any>
}

/**
 * Normalizes free-text food descriptions using deterministic parsing first, with AI fallback.
 */
export async function normalizeFoodDescription(
  raw_description: string
): Promise<FoodNormalizationResponse> {
  return apiClient<FoodNormalizationResponse>('/api/v1/ai/normalize-food', {
    method: 'POST',
    body: JSON.stringify({ raw_description }),
  })
}

/**
 * Retrieves AI-assisted suggestions for a surplus donation.
 * Note: Donor-declared quantity remains authoritative.
 */
export async function assistDonation(
  donation_id: string
): Promise<DonationAIAssistResponse> {
  return apiClient<DonationAIAssistResponse>(`/api/v1/donations/${donation_id}/ai-assist`, {
    method: 'POST',
  })
}

/**
 * Retrieves candidate suggestions for an NGO hunger relief need from free-text requirements.
 */
export async function assistNeed(
  need_id: string,
  free_text_requirement?: string
): Promise<NeedAIAssistResponse> {
  return apiClient<NeedAIAssistResponse>(`/api/v1/needs/${need_id}/ai-assist`, {
    method: 'POST',
    body: free_text_requirement ? JSON.stringify({ free_text_requirement }) : undefined,
  })
}

/**
 * Retrieves human-readable natural language explanation for a deterministic match priority score.
 */
export async function getMatchExplanation(
  match_id: string
): Promise<ExplanationResponse> {
  return apiClient<ExplanationResponse>(`/api/v1/matches/${match_id}/explanation`)
}

/**
 * Retrieves human-readable explanation for operational delivery status.
 */
export async function getDeliveryExplanation(
  delivery_id: string
): Promise<ExplanationResponse> {
  return apiClient<ExplanationResponse>(`/api/v1/deliveries/${delivery_id}/explanation`)
}

/**
 * Extracts candidate fields from verification document text/image and stores in verification_documents.extracted_data.
 */
export async function processVerificationDocument(
  document_id: string,
  ocr_text?: string,
  document_type?: string
): Promise<DocumentExtractionResult> {
  return apiClient<DocumentExtractionResult>(`/api/v1/documents/${document_id}/process-ocr`, {
    method: 'POST',
    body: JSON.stringify({
      document_id,
      ocr_text,
      document_type,
    }),
  })
}

/**
 * Performs deterministic consistency check between document candidates and registered entity/profile.
 */
export async function checkDocumentConsistency(
  document_id: string
): Promise<DocumentConsistencyResponse> {
  return apiClient<DocumentConsistencyResponse>(`/api/v1/documents/${document_id}/consistency-check`)
}

/**
 * Retrieves visual integrity analysis for a delivery (reusing Phase 14 integrity checks).
 */
export async function getIntegrityAnalysis(
  delivery_id: string
): Promise<any[]> {
  return apiClient<any[]>(`/api/v1/integrity/deliveries/${delivery_id}`)
}

/**
 * Retrieves human-readable environmental and meal impact explanation based on recorded data.
 */
export async function getImpactExplanation(
  entity_id: string,
  donation_id?: string
): Promise<ExplanationResponse> {
  const query = donation_id ? `?donation_id=${encodeURIComponent(donation_id)}` : ''
  return apiClient<ExplanationResponse>(`/api/v1/impact/${entity_id}/explanation${query}`)
}
